graphql_compiler/interpreter/debugging.py

```diff
@@ -0,0 +1,196 @@
+from copy import deepcopy
+from dataclasses import dataclass, field
+from typing import (
+    Any,
+    ClassVar,
+    Dict,
+    Generic,
+    Iterable,
+    List,
+    Sequence,
+    Tuple,
+    TypeVar,
+)
+
+from ..typedefs import Literal
+from .typedefs import DataToken
+
+
+T = TypeVar("T")
+U = TypeVar("U")
+
+
+def _enumerate_starting_at(
+    sequence: Sequence[T],
+    first_index: int,
+) -> Iterable[Tuple[int, T]]:
+    """Enumerate only a portion of a sequence, without slicing the sequence.
+
+    Equivalent to enumerate(sequence[first_index:], start=first_index). The built-in enumerate()
+    always iterates the entire sequence and only changes the counter's value based on
+    the start parameter. In this function, we skip elements with indices in [0, first_index)
+    in the sequence, and only iterate over elements with indices in [first_index, len(sequence)).
+
+    Args:
+        sequence: arbitrary sequence of values over which to iterate
+        first_index: the first index of the sequence to produce in the resulting iterable
+
+    Yields:
+        tuples (index, element_at_that_index_in_sequence), analogous to the built-in enumerate()
+    """
+    for index in range(first_index, len(sequence)):
+        yield index, sequence[index]
+
+
+def _unzip_and_yield_second(iterable: Iterable[Tuple[T, U]]) -> Iterable[U]:
+    """Unpack an iterable of tuples and only yield the second element of each tuple."""
+    for _, second in iterable:
+        yield second
+
+
+@dataclass(frozen=True)
+class AdapterOperation:
+    """The record of an action performed to or by an InterpreterAdapter function as part of a trace.
+
+    For example, calling get_tokens_of_type("Foo") would be an AdapterOperation with kind "call",
+    with name "get_tokens_of_type", with unique ID and parent unique IDs appropriate for the trace,
+    and with data (("Foo",), {}) as the tuple of the positional arguments tuple and kwargs dict.
+
+    The parent_uid field allows us to track dependencies across different operations. For example,
+    data yielded by a call to a generator-style function will be recorded in AdapterOperations
+    with kind "yield" whose parent_uid points to the AdapterOperation of the particular invocation
+    of the function that is yielding that data.
+    """
+
+    kind: Literal["call", "yield", "return"]
+    name: str
+    uid: int
+    parent_uid: int
+    data: Any
+
+
+@dataclass(frozen=True)
+class RecordedTrace(Generic[DataToken]):
+    """A complete immutable recording of the execution of a query via an InterpreterAdapter.
+
+    Includes a linearized sequence of all operations performed during the execution of the query:
+    - adapter function calls, yields, and returns;
+    - actuation of all input and output iterables.
```

Comment:
```suggestion
    Includes a linear sequence of all operations performed during the execution of the query.
```
- "Linearized" seems to suggest a transformation, but the sequence is actually just all the operations that occurred in their actual execution order on our single thread of execution, right?

- Isn't the "actuation of iterables" covered by the "adapter function yields"?

I don't think we need to enumerate all the operation types here, since we just covered them in the prior class's docstring and since the type annotations make it very clear where to look for additional information.

Reply:
It sort of is a transformation, in a sense. The trace represents one possible linearization of the execution, but there are many others that would also be valid and could be produced with a slightly different adapter implementation. "Linearized" here is (perhaps non-obviously) referring to the property of [linearizability](https://en.wikipedia.org/wiki/Linearizability) — the trace is a concrete linearization of the observed execution, a "proof by construction" that the interpreter satisfies the linearizability property. Linearizability isn't that important at the moment (the interpreter executes synchronously on a single thread) but becomes more interesting if/when we add `async` support.

For "actuation of iterables", you're right that it is partially covered by "adapter function yields", but it's not entirely the same — the `data_contexts` input argument to adapter functions is also actuated, but in that case the adapter function is _yielded to_. This operation is also recorded in the trace, and I felt it was worth mentioning here since the previous class' docstring wasn't explicit about that since it doesn't distinguish between the two kinds of yields.

This is all tricky, and I recognize that the current description doesn't do a good job of communicating what needs to be communicated. Let me know what you think about the above, and let's work together to find a better way to describe all this.

Reply:
Changed the "actuation of iterables" to make the two bullet points have no overlap between each other.